### PR TITLE
LOG4J2-2916 Avoid redundant Kafka producer instantiation causing thread leaks.

### PR DESCRIPTION
**Pull request by wuqian0808**
During reconfigure, LoggerContext will start new configuration and stop previous configuration. The firstime, prev is DefaultConfiguration, new configuration is eg: XmlConfiguration. Then the xml defined appender count in abstractmanager will be increased to 1. The nexttime reconfigure is called, xml defined appender count will be increased to 2. However the abstractmanager stop method will check if --count is <= 0. This cause the kafka producer thread never get stopped.

So each time loggercontext config start, the kafkamanager will create new kafka producer thread although it has one producer instance already. And the producer thread will never get closed.

The count is correct. It means the count of appenders using this manager. As we can see, the AbstarctManager will store a map of all static log4j2 managers. So during reconfigure, if managers is used by new config, the manager shouldn't be closed. Instead I just added a logic to check whether new kafka thread should be opended.

### CODE REVIEW

**Comment by wuqian0808** [comment]
Testcase added. Is there anything else I can do?

**Comment by wuqian0808** [comment]
@jvz  kindly help to advise

**Comment by jvz** [comment]
@wuqian0808 could you fix the filename of the test file? Without the .java extension, I'm pretty sure the compiler or Maven will complain.

**Comment by wuqian0808** [comment]
> @wuqian0808 could you fix the filename of the test file? Without the .java extension, I'm pretty sure the compiler or Maven will complain.

File name changed

**Review by jvz — APPROVED** [review]
Test failures appear to be unrelated. Looks good to me!

**Comment by wuqian0808** [comment]
@rgoers Could u help to advise? Or is there anything else should I do?

**Comment by wuqian0808hundsun** [comment]
@vy  Could u help to advise if there is anything else should I do before merging?

File: log4j-kafka/src/test/java/org/apache/logging/log4j/kafka/appender/KafkaThreadTest.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        int kafkaThreadCount = 0;
+        if (maps != null && maps.size() >0) {
+            for (Thread thread : maps.keySet()) {
+                if (thread.getName().startsWith("kafka-producer")){
+                    kafkaThreadCount++;
+                }
+            }
+        }
+        assertTrue(kafkaThreadCount <= 5);
+    }
```
Comment (vy) [inline log4j-kafka/src/test/java/org/apache/logging/log4j/kafka/appender/KafkaThreadTest.java:31]: Magical numbers in the tests, e.g., 5 here, have strong potential to bite us in the back later on. Would you mind revising this test as follows, please?

```java
@Test
void test_kafka_producer_threads_are_not_leaking_after_context_restart() {

    // Determine the initial number of threads.
    final LoggerContext context = ctx.getLoggerContext();
    final int initialThreadCount = kafkaProducerThreadCount();

    // Perform context restarts.
    final int contextRestartCount = 3;
    for (int i = 0; i < contextRestartCount; i++) {
        context.getContext(false).reconfigure();
    }

    // Verify the final thread count.
    final int lastThreadCount = kafkaProducerThreadCount();
    assertThat(lastThreadCount).equalsTo(initialThreadCount);

}

private static int kafkaProducerThreadCount() {
    // Return the number of threads prefixed with "kafka-producer".
}
```
Note that,
- This is sort of a pseudo-code I jotted in the comment box, hence you need to fill it in yourself.
- Try to use JUnit 5, if possible. (You can see examples in the source code.)

**Review by vy — CHANGES_REQUESTED** [review]
Looks really nice! Would you also mind updating `changes.xml` too, please?

**Comment by vy** [comment]
Nevermind, applied your changes manually. Will backport it to 2.x too. Thanks so much!

**Pull request closed without merging**